TESTS/mbed_drivers/sai/main.cpp

```diff
@@ -0,0 +1,293 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "utest/utest.h"
+#include "unity/unity.h"
+#include "greentea-client/test_env.h"
+#include "mbed.h"
+#include "math.h"
+
+#define THREAD_STACK_SIZE 1024
+
+#if !DEVICE_SAI
+#error [NOT_SUPPORTED] SAI not supported for this target
+#endif
+
+using namespace utest::v1;
+
+/* Since boards on CI do not have wire SAI loop-back connection, some of the functional
+ * tests can not be executed.
+ * If you want to fully test SAI support please connect pins as follows:
+ * - SAI_A_SD   <--> SAI_B_SD
+ * - SAI_A_BCLK <--> SAI_B_BCLK
+ * - SAI_A_WCLK <--> SAI_B_WCLK
+ * and enable loop-back tests by setting LOOPBACK_CONNECTION to true.
+ */
+#define LOOPBACK_CONNECTION (false)
+
+#define BUFFER_SIZE 100
+
+typedef enum
+{
+    LOOPBACK_TEST_OK, LOOPBACK_TEST_SKIP, LOOPBACK_TEST_FAILURE
+} loopback_test_tatus_t;
+
+static uint32_t transmit_buffer[BUFFER_SIZE];
+static uint32_t receive_buffer[BUFFER_SIZE];
+
+/* Auxiliary function which sets the specified buffer using specified pattern. */
+static void set_buffer(void * buffer, uint32_t size, char pattern)
+{
+    char* p_byte = (char*) buffer;
+
+    while (size--) {
+        *p_byte = pattern;
+        p_byte++;
+    }
+}
+
+/* Auxiliary function which compares two buffers and returns true if both are the same. */
+static bool compare_buffers(const void *buffer1, const void *buffer2, uint32_t size)
+{
+    const unsigned char *p1 = (const unsigned char *) buffer1;
+    const unsigned char *p2 = (const unsigned char *) buffer2;
+    while (size--) {
+        if (*p1 != *p2) {
+            return false;
+        } else {
+            p1++;
+            p2++;
+        }
+    }
+
+    return true;
+}
+
+/* Number of formats which must be supported by SAI device.
+ * This variable defines number of obligate formats in test_sai_formats array. */
+const uint32_t required_formats_count = 5;
```

Comment:
Minor: Is this used anywhere/needed? The array below only has a single element at the moment.

Reply:
I need to fix `test_sai_formats` formats array. While working on this test only one format was supported by the K64F, K66F example SAI driver, but according to the SAI requirements SAI device must support at least the following formats `sai_mode_i2s16, sai_mode_i2s16w32,  sai_mode_i2s32, sai_mode_pcm16l, sai_mode_pcm16s`.

Reply:
fixed